psbt: add non-default sighash types to PSBTs and unify sighash type match checking

achow101 opened the pull request:
Currently, we do not add the sighash field to PSBTs at all, even when we have signed with a non-default sighash. This PR changes the behavior such that when we (attempt to) sign with a sighash other than DEFAULT or ALL, the sighash type field will be added to the PSBT to inform the later signers that a different sighash type was used by a signer. Notably, this is necessary for MuSig2 support as all signers must sign using the same sighash type, but the sighash is not provided in partial signatures.

Furthermore, because the sighash type can also be provided on the command line, we require that if both a command line sighash type and the sighash field is present, they must specify the same sighash type. However, this was being checked by the wallet, rather than the signing code, so the `descriptorprocesspsbt` RPC was not enforcing this restriction at all, and in fact ignored the sighash field entirely. This PR refactors the checking code so that the underlying PSBT signing function `SignPSBTInput` does the check.

achow101 committed:
psbt: Return PSBTError from SignPSBTInput

SignPSBTInput will need to report the specific things that caused an
error to callers, so change it to return a PSBTError. Additionally some
callers will now check the return value and report an error to the user.

Currently, this should not change any behavior as the things that
SignPBSTInput will error on are all first checked by its callers.

achow101 committed:
psbt: Check sighash types in SignPSBTInput and take sighash as optional

achow101 committed:
tests: Test PSBT sighash type mismatch

achow101 committed:
rpc, psbt: Require sighashes match for descriptorprocesspsbt

achow101 committed:
script: Add IsPayToTaproot()

achow101 committed:
wallet: change FillPSBT to take sighash as optional

Instead of having the caller have to figure out the correct sane default
to provide to FillPSBT, have FillPSBT do that by having it take the
sighash type as an optional. This further allows it to distinguish
between an explicit sighash type being provided and expecting the
default value to be used.

achow101 committed:
wallet: Remove sighash type enforcement from FillPSBT

achow101 committed:
psbt: Add sighash types to PSBT when not DEFAULT or ALL

When an atypical sighash type is specified by the user, add it to the
PSBT so that further signing can enforce sighash type matching.

fanquake commented:
https://github.com/bitcoin/bitcoin/actions/runs/12700499877/job/35403281996?pr=31622#step:12:8987:
```bash
  test  2025-01-10T00:28:24.738000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "D:\a\bitcoin\bitcoin\build\test\functional\test_framework\test_framework.py", line 135, in main
                                       self.run_test()
                                     File "D:\a\bitcoin\bitcoin\build\test\functional\wallet_taproot.py", line 411, in run_test
                                       self.do_test(
                                     File "D:\a\bitcoin\bitcoin\build\test\functional\wallet_taproot.py", line 395, in do_test
                                       self.do_test_psbt(comment, pattern, privmap, treefn, keys[2*nkeys:3*nkeys], keys[3*nkeys:4*nkeys])
                                     File "D:\a\bitcoin\bitcoin\build\test\functional\wallet_taproot.py", line 363, in do_test_psbt
                                       assert "non_witness_utxo" not in psbtin
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                   AssertionError
 test  2025-01-10T00:28:24.738000Z TestFramework (DEBUG): Closing down network thread 
```

scgbckbone commented on src/psbt.cpp:
```diff
@@ -407,19 +410,38 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
+    if (input.sighash_type && input.sighash_type != sighash) {
+        return PSBTError::SIGHASH_MISMATCH;
+    } else {
+        if ((!utxo.scriptPubKey.IsPayToTaproot() && (sighash != SIGHASH_ALL && sighash != SIGHASH_DEFAULT)) ||
```
is there any specific reason to allow SIGHASH_DEFAULT on non-taproot utxo? Shouldn't it be error instead of rewriting sighash in `sign.cpp`

```
    // BASE/WITNESS_V0 signatures don't support explicit SIGHASH_DEFAULT, use SIGHASH_ALL instead.
    const int hashtype = nHashType == SIGHASH_DEFAULT ? SIGHASH_ALL : nHashType;
```

achow101 replied:
It's for compatibility, particularly when signing a transaction that has both taproot and non-taproot inputs.

Sjors commented on src/rpc/rawtransaction.cpp:
```diff
@@ -235,8 +235,10 @@ PartiallySignedTransaction ProcessPSBT(const std::string& psbt_string, const std
-        // As such, we ignore the return value as any errors just mean that we do not have enough information.
-        SignPSBTInput(provider, psbtx, /*index=*/i, &txdata, sighash_type, /*out_sigdata=*/nullptr, finalize);
+        // Only error for mismatching sighash types as it is critical that the sighash to sign with matches the PSBT's
+        if (SignPSBTInput(provider, psbtx, /*index=*/i, &txdata, sighash_type, /*out_sigdata=*/nullptr, finalize) == common::PSBTError::SIGHASH_MISMATCH) {
```
b70e83da4210e8098cbf3fc3b2289fc710828455: in case of a hardware wallet (external signer) I would want this to fail before any call to the device is made.

It's a bit hard to tell if that's the case, maybe this specific check can be refactored into its own function? That would make things easier to follow in general.

achow101 replied:
Nothing in this file has access to hardware signers. Those are wallet only.

Sjors commented on src/psbt.cpp:
```diff
@@ -372,13 +375,13 @@ PrecomputedTransactionData PrecomputePSBTData(const PartiallySignedTransaction&
 }
 
-bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, const PrecomputedTransactionData* txdata, int sighash,  SignatureData* out_sigdata, bool finalize)
+PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, const PrecomputedTransactionData* txdata, int sighash,  SignatureData* out_sigdata, bool finalize)
```
0970178f620acd695c49526d2b375eae489072c1: can we make this `nodiscard` and then explicitly ignore the result in `rpc/rawtransaction.cpp`?

achow101 replied:
Done

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -632,11 +632,12 @@ SigningResult LegacyScriptPubKeyMan::SignMessage(const std::string& message, con
     if (n_signed) {
         *n_signed = 0;
     }
+    if (!sighash_type) sighash_type = SIGHASH_DEFAULT;
```
e94bb4e5fc6dea67844a8495b948b679649b7b65: maybe use `SIGHASH_ALL` here, since the legacy wallet doesn't support Taproot.

achow101 replied:
This is temporary anyways and is removed in a later commit.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2532,11 +2533,12 @@ SigningResult DescriptorScriptPubKeyMan::SignMessage(const std::string& message,
     if (n_signed) {
         *n_signed = 0;
     }
+    if (!sighash_type) sighash_type = SIGHASH_DEFAULT;
```
e94bb4e5fc6dea67844a8495b948b679649b7b65: maybe do this inside the `psbtx.tx->vin` loop below and set `SIGHASH_ALL` or `SIGHASH_DEFAULT` depending on whether it's taproot.

And then replace the magic buried in `MutableTransactionSignatureCreator::CreateSig` with an `assert`.

achow101 replied:
This is temporary and will be done inside of `SignPSBTInput` in a later commit.

Sjors replied:
Do you mean in a followup PR? I'm still seeing this in `CreateSig` at the last commit of this PR:

```cpp
// BASE/WITNESS_V0 signatures don't support explicit SIGHASH_DEFAULT, use SIGHASH_ALL instead.
const int hashtype = nHashType == SIGHASH_DEFAULT ? SIGHASH_ALL : nHashType;
```

achow101 replied:
> Is that redundant after [34c33a7](https://github.com/bitcoin/bitcoin/commit/34c33a747128e1c26b1931ac3fa0c4ed6b6a7bb7)?

No. There are other non-PSBT callers of `ProduceSignature` which may still pass `SIGHASH_DEFAULT` to legacy input types, notably `signrawtranssaction`.

Sjors replied:
It might make things easier to follow if those callers pass the right flag as well. If it can be done without making this PR too huge.

Sjors replied:
In other words: for all functions that take a sighash argument, you either pass the correct sighash or nothing.

achow101 replied:
I don't think that's actually correct either as then passing SIGHASH_DEFAULT will be unable to sign for non-taproot at the same time.

It's also a much larger change that I think is out of scope for this PR.

Sjors replied:
> at the same time

You mean a transaction with both a taproot and non-taproot input?

achow101 replied:
> You mean a transaction with both a taproot and non-taproot input?

Yes

Sjors reviewed (commented):
Concept ACK

The `descriptorprocesspsbt` documentation for `sighashtype` currently says:

> The signature hash type to sign with if not specified by the PSBT.

Let's change that to:

> The signature hash type to sign with. If the PSBT already specifies a hash type, it must match. Otherwise it is added to the PSBT.

Some comments based on partial code review.

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -235,6 +241,22 @@ def test_sighash_mismatch(self):
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "SINGLE|ANYONECANPAY")
+
+        # No sighash type specified fails
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs)
```
The function that produces this specific error message does not have access to the arguments nor the PSBT to produce a specific message in each instance. Rewiring everything to allow that is a significant amount of work that I don't want to do in this PR.

achow101 committed:
psbt: use sighash type field to determine whether to remove non-witness utxos

Since the sighash type field is written for atypical sighash types, we
can look at that field to figure out whether the psbt contains
unnecessary transactions.

rkrux commented on src/psbt.cpp:
```diff
@@ -467,36 +467,41 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
     return sig_complete ? PSBTError::OK : PSBTError::INCOMPLETE;
```
Can this commit be a separate PR? It would be nice to have a functional test that checks non witness utxos are dropped if sighash type is not `SIGHASH_ANYONECANPAY`.  It can be added in this PR as well at the cost of a larger diff.

achow101 replied:
No, this was added because without it there were test failures.

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -237,6 +237,43 @@ def test_sighash_mismatch(self):
+
+        # Check that the PSBT has a sighash field on all inputs
+        dec_psbt = self.nodes[0].decodepsbt(psbt)
+        for input in dec_psbt["inputs"]:
+            assert_equal(input["sighash"], "ALL|ANYONECANPAY")
```
Atm, there is only 1 input in the PSBT. It would be nice to have another input that is Taproot and check for its sighash.

achow101 replied:
Done

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -235,6 +241,22 @@ def test_sighash_mismatch(self):
 
+        # Repeat with descriptorprocesspsbt
+        # Mismatching sighash type fails
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "DEFAULT")
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "ALL")
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "NONE")
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "SINGLE")
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "NONE|ANYONECANPAY")
+        assert_raises_rpc_error(-22, "Specified sighash value does not match value stored in PSBT", self.nodes[0].descriptorprocesspsbt, psbt, descs, "SINGLE|ANYONECANPAY")
```
Nit feel free to ignore: I ignored this duplication in the first commit but now that it's being done twice, a refactor could be to just iterate this in a loop of sighash types.

achow101 replied:
Done

rkrux reviewed (commented):
Concept ACK fefe0636d4ae7c246042276cacd60b22f5fc6bb9

Good PR with enough context to unpack for me. Left few comments, will review again soon.

Build and functional tests successful.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2545,11 +2540,6 @@ std::optional<PSBTError> DescriptorScriptPubKeyMan::FillPSBT(PartiallySignedTran
             continue;
         }
 
-        // Get the Sighash type
-        if (sign && input.sighash_type != std::nullopt && *input.sighash_type != sighash_type) {
-            return PSBTError::SIGHASH_MISMATCH;
-        }
-
```
Yes

achow101 commented on src/psbt.cpp:
```diff
@@ -423,6 +423,12 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
+        if ((!utxo.scriptPubKey.IsPayToTaproot() && (sighash != SIGHASH_ALL && sighash != SIGHASH_DEFAULT)) ||
+            (utxo.scriptPubKey.IsPayToTaproot() && sighash != SIGHASH_DEFAULT)
+        ) {
+            input.sighash_type = sighash;
```
Yes. Additionally, specifying `ALL` for taproot means adding another byte to the signature and also changes the sighash.

achow101 commented on src/psbt.cpp:
```diff
@@ -413,12 +413,25 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
     }
 
+    // Get the sighash type
+    // If both the field and the paramter are provided, they must match
```
Fixed

achow101 commented on src/psbt.cpp:
```diff
@@ -413,12 +413,25 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
+    // If both the field and the paramter are provided, they must match
+    // If only the psbt field is provided, refuse to sign. For user safety, the desired sighash must
+    // be provided if the PSBT wants something other than DEFAULT
+    // If only the paramter is provider, use it and add it to the PSBT if it is other than SIGHASH_DEFAULT
```
Fixed

achow101 commented on src/wallet/external_signer_scriptpubkeyman.h:
```diff
@@ -35,7 +35,7 @@ class ExternalSignerScriptPubKeyMan : public DescriptorScriptPubKeyMan
  util::Result<void> DisplayAddress(const CTxDestination& dest, const ExternalSigner& signer) const;
 
-  std::optional<common::PSBTError> FillPSBT(PartiallySignedTransaction& psbt, const PrecomputedTransactionData& txdata, int sighash_type = 1 /* SIGHASH_ALL */, bool sign = true, bool bip32derivs = false, int* n_signed = nullptr, bool finalize = true) const override;
+  std::optional<common::PSBTError> FillPSBT(PartiallySignedTransaction& psbt, const PrecomputedTransactionData& txdata, std::optional<int> sighash_type = 1 /* SIGHASH_ALL */, bool sign = true, bool bip32derivs = false, int* n_signed = nullptr, bool finalize = true) const override;
```
Indeed, Fixed

achow101 commented on src/psbt.cpp:
```diff
@@ -413,12 +413,25 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
 
+    // Get the sighash type
+    // If both the field and the paramter are provided, they must match
+    // If only the psbt field is provided, refuse to sign. For user safety, the desired sighash must
```
I've removed this sentence since we allow `DEFAULT` and `ALL`.

fjahr commented on src/common/messages.cpp:
```diff
@@ -116,6 +116,10 @@ bilingual_str PSBTErrorString(PSBTError err)
         case PSBTError::UNSUPPORTED:
             return Untranslated("Signer does not support PSBT");
+        case PSBTError::INCOMPLETE:
+            return Untranslated("Input needs additional signatures or other data");
```
It looks like there is no test coverage for this yet. Is there coverage for this in a follow-up PR (maybe Musig2) or are you writing a test? Otherwise I can give it a shot.

achow101 replied:
I don't think there's any specific coverage for this, feel free to tackle it.

marshallshen replied:
Follow up: I created a PR to add test coverage: https://github.com/bitcoin/bitcoin/pull/33025
Will ensure it passes all the CI checks, 🙇

fjahr reviewed (commented):
Concept ACK

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -230,6 +230,47 @@ def test_sighash_mismatch(self):
+        assert_equal(fin_res["complete"], True)
+        fin_hex = fin_res["hex"]
+
+        # Change the sighash field to a different value and make sure we still finalize to the same thing
+        mod_psbt = PSBT.from_base64(psbt)
+        mod_psbt.i[0].map[PSBT_IN_SIGHASH_TYPE] = (SIGHASH_ALL).to_bytes(4, byteorder="little")
+        if self.options.descriptors:
+            mod_psbt.i[1].map[PSBT_IN_SIGHASH_TYPE] = (SIGHASH_ALL).to_bytes(4, byteorder="little")
+        psbt = mod_psbt.to_base64()
+        fin_res = self.nodes[0].finalizepsbt(psbt)
+        assert_equal(fin_res["complete"], True)
+        assert_equal(fin_res["hex"], fin_hex)
```
This is the final hex when I run the test, which is same as the one before modifying the `sighash` type as per the test:
```
0200000000010208bd94fdd97de6bc3ec454b8e0088a5341f9e9611bfbe412387e7e0df35207960000000000fdffffff08bd94fdd97de6bc3ec454b8e0088a5341f9e9611bfbe412387e7e0df35207960100000000fdffffff0280f0fa0200000000160014f57919cd765629cd0d2ba39021a146ee065d697508c2f00800000000160014c7109263e9f5db5d2162ce40a1f898f6ac3933ad02473044022067caa211634d7f25bcac48704165ca7d257ff80a7d0b764825877c7a4a137eff02204ef676f27c9f47813be48a0d2cce9b1dfa2b6fb0ac44eb1d4273781299ad318c81210362cd5a24c1978e204d78b15e5b303c4a8bbb8234591d94b1b8faa27cc72fdeed01412427cd2bcb8d07bc16ccb73a4cffeb996f89a6ee80a39d8ca4cf80f342039e0ec90d12dd0bcb1cdc392c05b1b9e2006fb23fb5fb9ade3a30f9c7b40cd3a581a78100000000
```

These are the 2 signatures of the 2 inputs:
```
3044022067caa211634d7f25bcac48704165ca7d257ff80a7d0b764825877c7a4a137eff02204ef676f27c9f47813be48a0d2cce9b1dfa2b6fb0ac44eb1d4273781299ad318c81
2427cd2bcb8d07bc16ccb73a4cffeb996f89a6ee80a39d8ca4cf80f342039e0ec90d12dd0bcb1cdc392c05b1b9e2006fb23fb5fb9ade3a30f9c7b40cd3a581a781
```

I notice the `81` at the end in both of them signifying `ALL|ANYONECANPAY`.
Why does it not change to `01` when the transaction inputs are signed using `SIGHASH_ALL` later, and subsequently leading to changing of actual signature as well because all inputs are signed now?
Is it because the `finalizepsbt` doesn't actually sign the inputs again because there are signatures already in the PSBT albeit using a different sighash type? Shouldn't `PSBTInputSignedAndVerified()` fail here because the sighash type doesn't match with the signature?

achow101 replied:
> Is it because the `finalizepsbt` doesn't actually sign the inputs again because there are signatures already in the PSBT albeit using a different sighash type?

`finalizepsbt` does not actually sign even though it calls the sign functions. It does not have access to private keys.

> Shouldn't `PSBTInputSignedAndVerified()` fail here because the sighash type doesn't match with the signature?

The BIP states that the sighash field is merely a suggestion and signers are allowed to ignore it. Thus signatures can be produced with a different sighash.

rkrux replied:
Thanks, I get it now. I asked because I noticed `SignPSBTInput` being called but I see now the `DUMMY_SIGNING_PROVIDER` argument passed to it as well.
I see the reason why the test is contrived such that it requires manually updating the sighash and calling `finalisepsbt` again.

rkrux replied:
> The BIP states that the sighash field is merely a suggestion

Can you share which BIP is this that states it's a suggestion? I found a contradicting statement in [BIP 174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki), which is confusing.

> Signatures for this input must use the sighash type, finalizers must fail to finalize inputs which have signatures that do not match the specified sighash type.

achow101 replied:
Huh, it does say that. Perhaps I'm remembering an old draft where it was different.

Changed the test to check that it doesn't finalize, and added a commit to implement that behavior.

rkrux commented on src/psbt.cpp:
```diff
@@ -407,19 +410,37 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
+    // Get the sighash type
+    // If both the field and the parameter are provided, they must match
+    // For user safety, the desired sighash must be provided if the PSBT wants something other than DEFAULT
+    // If only the parameter is provider, use it and add it to the PSBT if it is other than SIGHASH_DEFAULT
```
`provided`

achow101 replied:
Done

rkrux reviewed (commented):
I reviewed range-diff, asked a question.

```
git range-diff fefe0636d4ae7c246042276cacd60b22f5fc6bb9...c99e1bc359263d44a85460032c04f7c1f3c688c7
```

All the new changes address the previous review comments, namely: removing duplication in the functional test rpc_psbt, adding Taproot input in the test, fixing comments, and changing `FillPSBT()` argument default.

Sjors commented:
Silent merge conflict with bitcoin-core/gui#850.

rkrux reviewed (commented):
ACK 9911ba82be9ec2243b86bb149d6ec8509d9717d7

I reviewed the range diff this time, my earlier query is answered.

```
git range-diff c99e1bc359263d44a85460032c04f7c1f3c688c7...9911ba82be9ec2243b86bb149d6ec8509d9717d7
```

Sjors reviewed (commented):
Reviewed code for 9911ba82be9ec2243b86bb149d6ec8509d9717d7, but I still need to review the last two commits.

`rpc_psbt.py --descriptors` consistently fails for me in commit 121c8fb494df4e81566eb0706f7b9500a870ac85 "psbt: Check sighash types in SignPSBTInput and take sighash as optional", at "Check that non-witness UTXOs are removed for segwit v1+ inputs" with `assert not non_witness_utxo` at line 136.

But only on Ubuntu, not on macOS, so maybe a gcc vs clang thing?

It's happy again three commits later at 6ec298a335585ef723d113b94959377d3ff65759 "psbt: use sighash type field to determine whether to remove non-witness utxos"

Sjors commented on src/psbt.cpp:
```diff
@@ -422,6 +422,12 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
     if (!sighash) sighash = utxo.scriptPubKey.IsPayToTaproot() ? SIGHASH_DEFAULT : SIGHASH_ALL;
     if (input.sighash_type && input.sighash_type != sighash) {
         return PSBTError::SIGHASH_MISMATCH;
+    } else {
```
(although maybe just indenting the second line with one extra space would do the trick

achow101 replied:
This is all explained by the comment 2 lines up.

Took the suggestion, changed the comment to be less repetitive of things already said.

Sjors replied:
In 9917a47ca49456172b03fee1e75c79ee72c381a6 "psbt: Add sighash types to PSBT when not DEFAULT or ALL"

Better. Maybe also drop the `else`, since the `SIGHASH_MISMATCH` guard makes sense isolation.

achow101 replied:
Removed the `else`.

Sjors commented on src/psbt.cpp:
```diff
@@ -466,36 +466,41 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
+            to_drop.clear();
+            break;
+        }
+        // non_witness_utxos can only be dropped if the sighash type does not include SIGHASH_ANYONECANPAY
```
In 6ec298a335585ef723d113b94959377d3ff65759 "psbt: use sighash type field to determine whether to remove non-witness utxos":

"can only ... if not" is hard to parse, and it seems the code doing the opposite. It also seems to do the opposite of the original behavior.

```
// Only drop non_witness_utxos if sighash_type != SIGHASH_ANYONECANPAY
```

Can you clarify the reason why `SIGHASH_ANYONECANPAY` requires different treatment in the first place?

achow101 replied:
Fixed and reworded the comment.

`SIGHASH_ANYONECANPAY` does not commit to any input information, so none of the protections that allow us to drop the `non_witness_utxo` apply.

Sjors replied:
Ok, in 795e911ded2563705cec69e43ff622ace22d58d5 "psbt: use sighash type field to determine whether to remove non-witness utxos" it's now more clear that this just moves a check inside the `for` loop.

I still can't wrap my head around the logic of how and why we drop `non_witness_utxos`.

Sjors commented on src/psbt.cpp:
```diff
@@ -466,36 +466,41 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
+        int wit_ver;
+        std::vector<unsigned char> wit_prog;
+        if (input.witness_utxo.IsNull() || !input.witness_utxo.scriptPubKey.IsWitnessProgram(wit_ver, wit_prog)) {
+            // There's a non-segwit input or Segwit v0, so we cannot drop any witness_utxos
```
In https://github.com/bitcoin/bitcoin/commit/6ec298a335585ef723d113b94959377d3ff65759 "psbt: use sighash type field to determine whether to remove non-witness utxos": any non_witness_utxos? Also we don't know if it's SegWit v0 at this point?

achow101 replied:
Fixed

Sjors reviewed (commented):
Also looked at the last two commits.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -665,7 +665,10 @@ std::optional<PSBTError> LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransact
         }
-        SignPSBTInput(HidingSigningProvider(this, !sign, !bip32derivs), psbtx, i, &txdata, sighash_type, nullptr, finalize);
+        PSBTError res = SignPSBTInput(HidingSigningProvider(this, !sign, !bip32derivs), psbtx, i, &txdata, sighash_type, nullptr, finalize);
+        if (res != PSBTError::OK && res != PSBTError::INCOMPLETE) {
```
Updated the commit message.

It's not a problem before, and not a problem now, because all of the things `SignPSBTInput` would fail on are checked by its callers.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2202,7 +2202,6 @@ std::optional<PSBTError> CWallet::FillPSBT(PartiallySignedTransaction& psbtx, bo
     if (n_signed) {
         *n_signed = 0;
     }
-    if (!sighash_type) sighash_type = SIGHASH_DEFAULT;
```
Changed `RemoveUnnecessaryTransactions` to take sighash type as optional in this commit as well.

achow101 commented on src/psbt.cpp:
```diff
@@ -413,12 +413,24 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
 
+    // Get the sighash type
+    // If both the field and the parameter are provided, they must match
+    // For user safety, the desired sighash must be provided if the PSBT wants something other than DEFAULT
```
Done

Sjors commented:
Code review ACK 61885ad9d406e25ec3b6523d01918e886996f1c3

yancyribbens commented on src/psbt.cpp:
```diff
@@ -407,19 +410,38 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
     }
 
+    // Get the sighash type
+    // If both the field and the parameter are provided, they must match
```
nit missing period after match

yancyribbens commented on src/psbt.cpp:
```diff
@@ -407,19 +410,38 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
+    if (input.sighash_type && input.sighash_type != sighash) {
+        return PSBTError::SIGHASH_MISMATCH;
+    }
+    // Set the PSBT sighash field when sighash is not DEFAULT or ALL
```
nit period after `ALL`

Sjors commented:
re-ACK 61885ad9d406e25ec3b6523d01918e886996f1c3

Dropped the else: https://github.com/bitcoin/bitcoin/pull/31622#discussion_r1963881147

rkrux commented:
> re-ACK [61885ad](https://github.com/bitcoin/bitcoin/commit/61885ad9d406e25ec3b6523d01918e886996f1c3)
> 
> Dropped the else: [#31622 (comment)](https://github.com/bitcoin/bitcoin/pull/31622#discussion_r1963881147)

@Sjors 61885ad9d406e25ec3b6523d01918e886996f1c3 commit seems to be outdated now.

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -230,6 +230,47 @@ def test_sighash_mismatch(self):
+        assert_equal(fin_res["complete"], True)
+        fin_hex = fin_res["hex"]
+
+        # Change the sighash field to a different value and make sure we still finalize to the same thing
```
What scenario is not tested if this flow is not done and the transaction generated above is mined directly?

achow101 replied:
Nothing, it just helps to mutate one which we know is definitely good.

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -230,6 +230,47 @@ def test_sighash_mismatch(self):
+
+        # Make a PSBT
+        psbt = wallet.walletcreatefundedpsbt(wallet.listunspent(), [{def_wallet.getnewaddress(): 0.5}])["psbt"]
+        psbt = wallet.walletprocesspsbt(psbt=psbt, sighashtype="ALL|ANYONECANPAY", finalize=False)["psbt"]
```
The `descriptorprocesspsbt` RPC uses a different flow than the `walletprocesspsbt` RPC because it calls `ProcessPSBT()` instead of `FillPSBT()`. I think it'd be nice if this tests also ensures the sighash is added via the `descriptorprocesspsbt` RPC as well.

achow101 replied:
Done

rkrux commented on src/psbt.cpp:
```diff
@@ -467,37 +467,41 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
 }
 
-void RemoveUnnecessaryTransactions(PartiallySignedTransaction& psbtx, std::optional<int> sighash_type)
+void RemoveUnnecessaryTransactions(PartiallySignedTransaction& psbtx)
```
Re: https://github.com/bitcoin/bitcoin/pull/31622#discussion_r1933698058

Earlier I thought we should add a test that checks the removal of `non_witness_utxo` in case of segwit v1 inputs only but I see there is one already here `test_utxo_conversion` in `rpc_psbt`. However, it works using `walletprocesspsbt` RPC only. I think this PR can test for the same functionality via `descriptorprocesspsbt` RPC too because it uses a different flow internally. Either of `test_utxo_conversion` or `test_sighash_adding` can be updated for this.

achow101 replied:
That's orthogonal to this PR and can be done separately.

rkrux replied:
I attempted this^ in PR #32413.

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -230,6 +230,47 @@ def test_sighash_mismatch(self):
 
         wallet.unloadwallet()
 
+    def test_sighash_adding(self):
```
Let's add a TODO in this file to write a test that checks `PSBT_GLOBAL_TX_MODIFIABLE` field is correctly set if `SIGHASH_SINGLE` is used in any of the inputs once the PSBT V2 PR https://github.com/bitcoin/bitcoin/pull/21283 is merged?

achow101 replied:
I don't think it makes sense to do that here since PSBTv2 is still entirely unparseable.

rkrux replied:
I know it's not parseable right now, that's why suggested only a TODO that'd be done after PSBTv2 gets merged. The assumption I have is this PR will get merged before #21283. I didn't notice any such test in the PSBT PR, which is fine because that PR is already big enough, and adding a TODO here can be a reminder for later.

rkrux reviewed (commented):
Few comments.

achow101 committed:
psbt: Enforce sighash type of signatures matches psbt

BIP 174 states that the sighash type of all signatures must match the
type given by the PSBT, so do that.

Sjors commented:
> From what I understand, it's for efficiency.

I understand the general reason, but the code itself and the nuance for different sighash flags is hard to follow.

achow101 commented:
The fuzzer CI tasks are failing since the change in sighash type enforcement. Presumably this is because the fuzz corpus contains seeds that generate psbts which violate the new constraint. How do we move forward with resolving that?

dergoegge commented on src/psbt.cpp:
```diff
@@ -407,19 +410,58 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
+            if (sig.back() != *sighash) return PSBTError::SIGHASH_MISMATCH;
+        }
+        for (const auto& [_, sig] : input.partial_sigs) {
+            if (sig.second.back() != *sighash) return PSBTError::SIGHASH_MISMATCH;
```
The fuzz test decodes a psbt from the fuzz input and passes it to the various psbt functions. Looks like it crashes here because the signature is empty? (could be wrong, just glanced at this)

achow101 replied:
Fixed

achow101 committed:
psbt: Require ECDSA signatures to be validly encoded

Needed for later validation of sighash types.

achow101 commented:
Added a commit that should fix the fuzz failure.

rkrux commented:
The functional test failure seems unrelated to me. A re-run should fix it?

`test_preferred_tiebreaker_inv()` in `p2p_tx_download.py`

rkrux commented on test/functional/rpc_psbt.py:
```diff
@@ -201,6 +202,100 @@ def test_input_confs_control(self):
+            assert_equal(fin_res["complete"], True)
+            fin_hex = fin_res["hex"]
+
+            # Change the sighash field to a different value and make sure we can no longer finalize
```
Nice to see this now, it was bugging me earlier (but could not prove to myself until I found it in the spec) that how it can finalise to the same hex with a different sighash when underneath `PSBTInputSignedAndVerified` uses the same `interpreter.cpp::VerifyScript()` that is used in other script validation flows as well.

theStack reviewed (commented):
Concept ACK

rkrux commented on src/psbt.h:
```diff
@@ -429,6 +429,11 @@ struct PSBTInput
                     s >> sig;
 
+                    // Check that the signature is validly encoded
+                    if (sig.empty() || !CheckSignatureEncoding(sig, SCRIPT_VERIFY_DERSIG | SCRIPT_VERIFY_STRICTENC, nullptr)) {
```
> nit: CheckSignatureEncoding already checks for empty sig, so the first check isn't needed

I checked that `CheckSignatureEncoding` returns `true` for an empty signature.
https://github.com/bitcoin/bitcoin/blob/ff69046e664f2460bb2ef595c5bd16eeb4a13f58/src/script/interpreter.cpp#L200-L205

But here the requirement is to throw error in that case because of what's stated in [BIP 174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki).

> The signature as would be pushed to the stack from a scriptSig or witness. The signature should be a valid ECDSA signature corresponding to the pubkey that would return true when verified and not a value that would return false or be invalid otherwise (such as a NULLDUMMY).

Just like it's handled in `CheckECDSASignature`:
https://github.com/bitcoin/bitcoin/blob/ff69046e664f2460bb2ef595c5bd16eeb4a13f58/src/script/interpreter.cpp#L1655-L1657

Or in `VerifyECDSASignature` via `CPubKey::Verify`:
https://github.com/bitcoin/bitcoin/blob/971952588daebf0d4d08a7e0eeef1734fcd8a9b1/src/pubkey.cpp#L56-L58

That's why I believe `sig.empty()` is added separately.

rkrux replied:
> any reason why the verification flags don't also include SCRIPT_VERIFY_LOW_S?

I had been wondering about this as well. I think this flag should be included as the core wallet always produces low R/S values ECDSA signatures already ever since this change: #13666. 

If I am not missing anything, might as well the wallet reject such PSBTs while unserializing?

rkrux replied:
Also, I realise that it would be cool to pass a reference to the last argument to `CheckSignatureEncoding` that captures the error so that we can relay it forward to the user; the current error message (or its variation) can remain as a prefix.

achow101 replied:
> That's why I believe `sig.empty()` is added separately.

Yes, that's why.

> any reason why the verification flags don't also include `SCRIPT_VERIFY_LOW_S`?

Low S is a standardness rule, not a consensus rule. PSBTs may be produced by software other than Bitcoin Core, and these may sign with high S, for whatever reason. Since these are consensus valid signatures, we need to also accept them when decoding a PSBT.

rkrux commented:
I'm reviewing the PR again and will share a final review in the coming days.

rkrux commented on src/psbt.h:
```diff
@@ -429,6 +429,11 @@ struct PSBTInput
                     std::vector<unsigned char> sig;
```
Note for this commit: 14e35e675c576c770641160e982713dda4032bd8

I _think_ this is added because later in the commit 94fb539f4e22ed4ec9a65f90db26cb81e5b8749a, the last element of the partial sig is used to check for the sighash type match while signing the PSBT input for which this signature validity guarantee is needed. 

It'd be nice to mention something around this in the commit message to aid review (& for future reference).

achow101 replied:
Added a sentence to the commit message.

rkrux reviewed (commented):
Code review a7f76b28a1293929d05f15e0a9276d4ec3687c92.
I reviewed focussing on new commits that are added and will do a subsequent review that finalises my understanding of the whole diff.

achow101 commented on src/psbt.cpp:
```diff
@@ -423,6 +423,13 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
+                                            (sighash != SIGHASH_DEFAULT && sighash != SIGHASH_ALL)) {
+        input.sighash_type = sighash;
+    }
     Assert(sighash.has_value());
```
Done

achow101 commented on src/psbt.cpp:
```diff
@@ -425,6 +425,26 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
+        }
+        for (const auto& [_, sig] : input.m_tap_script_sigs) {
+            if (sig.back() != *sighash) return PSBTError::SIGHASH_MISMATCH;
+        }
```
I've added size checks. I did not gate behind a `IsPayToTaproot` as the fields should only be set if the output script is taproot anyways.

achow101 commented on src/psbt.cpp:
```diff
@@ -425,6 +425,26 @@ PSBTError SignPSBTInput(const SigningProvider& provider, PartiallySignedTransact
 
+    // Check all existing signatures use the sighash type
+    if (sighash == SIGHASH_DEFAULT) {
+        if (!input.m_tap_key_sig.empty() && input.m_tap_key_sig.size() != 64) {
+            return PSBTError::SIGHASH_MISMATCH;
+        }
+        for (const auto& [_, sig] : input.m_tap_script_sigs) {
+            if (sig.size() != 64) return PSBTError::SIGHASH_MISMATCH;
+        }
```
Meh, maybe someone can do that in a followup.

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -201,6 +202,34 @@ def test_input_confs_control(self):
         wallet.unloadwallet()
 
+    def test_sighash_mismatch(self):
+        self.log.info("Test sighash type mismatches")
```
I don't think it's really necessary as these log lines are in the PSBT test, and the only way to sign with PSBTs is with `*processpsbt`.

rkrux commented on src/psbt.cpp:
```diff
@@ -442,39 +484,44 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
-void RemoveUnnecessaryTransactions(PartiallySignedTransaction& psbtx, const int& sighash_type)
+void RemoveUnnecessaryTransactions(PartiallySignedTransaction& psbtx)
 {
-    // Only drop non_witness_utxos if sighash_type != SIGHASH_ANYONECANPAY
-    if ((sighash_type & 0x80) != SIGHASH_ANYONECANPAY) {
-        // Figure out if any non_witness_utxos should be dropped
-        std::vector<unsigned int> to_drop;
-        for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
```
It appears that the logic of the removal of unnecessary transactions has become more fine-grained as it checks the `sighash` of each input now instead of treating all the inputs with the same sighash as done previously.

Overall seems better to me than it was earlier, I've not thought of any adverse consequence of it yet.

rkrux reviewed (commented):
I'm done with review ea751ce.

The `wallet_backwards_compatibility.py` test is failing. Guessing it's because of the silent merge conflict issue #32369, I don't see the exact error though in the Cirrus CI.

I will ack once the CI is green.

achow101 commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -83,7 +83,7 @@ void PSBTOperationsDialog::signTransaction()
 
     WalletModel::UnlockContext ctx(m_wallet_model->requestUnlock());
 
-    const auto err{m_wallet_model->wallet().fillPSBT(SIGHASH_DEFAULT, /*sign=*/true, /*bip32derivs=*/true, &n_signed, m_transaction_data, complete)};
```
DEFAULT becomes ALL for non-taproot. It's just easier to put DEFAULT everywhere since it covers all cases.

rkrux reviewed (approved):
tACK b6101b1849fbc2ec510c56cfae39d7de244a50cc

I found this pull more involved because of multiple moving pieces and I _think_ I have covered all the areas that I could. Hence, sharing below my own notes of the review that hopefully would be helpful for anyone else reviewing the PR (& also my future self).

I see the main changes are along 3 lines:
1. Addition of atypical (non default/all) sighash type in the PSBT (Input).
2. Checking of the sighash type in the PSBT against the one provided in the RPC/CLI such as `descriptorprocesspsbt`, `walletprocesspsbt`.
3. Checking of all the different types of signatures in the PSBT inputs against the sighash type present in the PSBT input as per [BIP 174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki); it would be helpful to mention in the PR description.

To achieve the above changes, `SignPSBTInput` is refactored to accommodate the above changes, and now returns a `PSBTError` instead of a boolean that can't be discarded by the callers (enforced in compile time) and therefore must handle it.

Since `SignPSBTInput` is a common function used in multiple places & now checks for sighash as well (which is not always determined by the callers), `SignPSBTInput` now takes sighash as an optional leading to all the call sites passing in an optional if not being done already. This is what causes the diff to increase a bit as the call sites such as `CWallet::FillPSBT -> ScriptPubKeyMan::FillPSBT` & `ProcessPSBT` need to be updated along with their own callers as well that are many & are listed below.

```
FillPSBT callers -> walletprocesspsbt; walletcreatefundedpsbt; feebumper; ExternalSignerScriptPubKeyMan::FillPSBT; interfaces::wallet::fillPSBT (includes the GUI models)

ProcessPSBT callers -> descriptorprocesspsbt; utxoupdatepsbt
``` 

Another consequence of pushing down the sighash check in the common  `SignPSBTInput` is that the `descriptorprocesspsbt` RPC also rightly handles it now that was not being done earlier as this check was part of the `CWallet::FillPSBT` flow, which is separate from the `ProcessPSBT` flow; this is tested in the `sighash_mismatch` subtest in the `rpc_psbt` functional test. Naturally, this check from the `CWallet::FillPSBT` is removed now to avoid duplication.

One more consequence is that the `FinalisePSBT` RPC also accounts for the sighash check because it internally uses `SignPSBTInput` and it's used while verifying in the `sighash_adding` functional test. This conforms with what's mentioned in BIP 174.

Also to verify whether PSBT input signatures actually conform with the PSBT input sighash type, the encoding of the ECDSA signatures is checked while unserialising the PSBT. Notably, PSBTs with high S signatures are allowed to make the core wallet compatible with other signers - fine because these though are non-standard, but not consensus-invalid.

Lastly, the removal of non-witness utxos from the PSBT in case of only segwit v1 inputs is refactored to take into account the newly added sighash type in the PSBT input as opposed to using the one that was passed in from the caller, leading to a more granular checking imo; this particular flow was already tested in `test_utxo_conversion` functional subtest.

theStack commented on test/functional/rpc_psbt.py:
```diff
@@ -228,6 +228,53 @@ def test_sighash_mismatch(self):
+            fin_res = self.nodes[0].finalizepsbt(psbt)
+            assert_equal(fin_res["complete"], False)
+
+        self.nodes[0].sendrawtransaction(fin_hex)
+        self.generate(self.nodes[0], 1)
```
Rather than sending and mining only one finalized transaction, could instead use `testmempoolaccept` within the loop above, so the full check if the transaction is valid can be done for both the `wallet_psbt` and `desc_psbt` result (currently it's only done for the latter; the first `fin_hex` result is overwritten).

theStack commented on src/wallet/feebumper.cpp:
```diff
@@ -338,8 +338,8 @@ bool SignTransaction(CWallet& wallet, CMutableTransaction& mtx) {
         bool complete;
-        wallet.FillPSBT(psbtx, complete, SIGHASH_ALL, false /* sign */, true /* bip32derivs */);
-        auto err{wallet.FillPSBT(psbtx, complete, SIGHASH_ALL, true /* sign */, false  /* bip32derivs */)};
+        wallet.FillPSBT(psbtx, complete, std::nullopt, false /* sign */, true /* bip32derivs */);
+        auto err{wallet.FillPSBT(psbtx, complete, std::nullopt, true /* sign */, false  /* bip32derivs */)};
```
nit: while touching these lines anyway, could fix the extra-weird formatting of the named argument comments.

theStack commented on test/functional/rpc_psbt.py:
```diff
@@ -228,6 +228,53 @@ def test_sighash_mismatch(self):
+        wallet = self.nodes[0].get_wallet_rpc("sighash_adding")
+        def_wallet = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+
+        addr = wallet.getnewaddress(address_type="bech32")
+        outputs = [{addr: 1}]
+        outputs.append({wallet.getnewaddress(address_type="bech32m"): 1})
```
nit: `addr` is never used anywhere else, could directly construct the `outputs` object in one go

achow101 replied:
Done

theStack reviewed (approved):
Code-review ACK b6101b1849fbc2ec510c56cfae39d7de244a50cc

Left some nits below, nothing critical. As others have already noted, some more test coverage for certain areas would be nice, but can be done in a follow-up, e.g.:
* valid signature encoding check in parsing of `PSBT_IN_PARTIAL_SIG` (in particular, that a [high-s signature is accepted](https://github.com/bitcoin/bitcoin/pull/31622#discussion_r2064644621))
* [`PSBTError::INCOMPLETE`](https://github.com/bitcoin/bitcoin/pull/31622#discussion_r1938366105) error condition
* unit test for `CScript::IsPayToTaproot` (trivial)

rkrux reviewed (approved):
re-ACK f09e1f9

```
git range-diff b6101b1...f09e1f9
```
Range diff contains changes from [this review](https://github.com/bitcoin/bitcoin/pull/31622#pullrequestreview-2818667962) & rebase now that #28710 is merged that has led to the removal of `LegacyScriptPubKeyMan` and naturally all the changes in it from this PR - the overall diff has become marginally smaller.

theStack commented on test/functional/rpc_psbt.py:
```diff
@@ -228,6 +228,53 @@ def test_sighash_mismatch(self):
+            fin_res = self.nodes[0].finalizepsbt(psbt)
+            assert_equal(fin_res["complete"], False)
+
+        self.nodes[0].sendrawtransaction(fin_hex)
+        self.generate(self.nodes[0], 1)
```
nitty nit: now that the checks are done via `testmempool` above, these two lines could be deleted

rkrux replied:
I feel it's fine to keep them - imho the ultimate test is whether the transaction can be mined, gives me more confidence.

theStack reviewed (approved):
re-ACK f09e1f943dbd1e5865dece3c35f342e4f0b31b9e :writing_hand: :hash:

theStack reviewed (approved):
re-ACK ee045b61efc1479c1866b786661ef39a863677d0
(as per `$ git range-diff f09e1f94...ee045b61`)

rkrux reviewed (approved):
re-ACK ee045b61efc1479c1866b786661ef39a863677d0

Only rebase changes from #32305 
```
git range-diff f09e1f94...ee045b61
```

fjahr commented on src/common/types.h:
```diff
@@ -20,6 +20,8 @@ enum class PSBTError {
     EXTERNAL_SIGNER_FAILED,
     UNSUPPORTED,
+    INCOMPLETE,
+    OK,
```
nit: Having an `OK` variant makes this more into an `PSBTResult` rather than `PSBTError`

achow101 replied:
Will leave for a followup

fjahr commented on src/interfaces/wallet.h:
```diff
@@ -204,7 +204,7 @@ class Wallet
 
     //! Fill PSBT.
-    virtual std::optional<common::PSBTError> fillPSBT(int sighash_type,
+    virtual std::optional<common::PSBTError> fillPSBT(std::optional<int> sighash_type,
```
Since there is now a `PSBTError::OK` it seems like there is no need for using an `std::optional<common::PSBTError>` return type. But this can be made more consistent in a follow-up.

achow101 replied:
Will leave for a followup

fjahr reviewed (commented):
Code review ACK ee045b61efc1479c1866b786661ef39a863677d0

My suggestions make more sense as a follow-up than addressing them here.
